File: 01_arrays/d04_squares_of_a_sorted_array.go
package main

import (
	"fmt"
)

func abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

func sortedSquares(nums []int) []int {
	res := make([]int, len(nums))
	left, right := 0, len(nums)-1
	for i := len(nums) - 1; i >= 0; i-- {
		if abs(nums[left]) < abs(nums[right]) {
			res[i] = nums[right] * nums[right]
			right--
		} else {
			res[i] = nums[left] * nums[left]
			left++
		}
	}

	return res
}

func main() {
	testNums := []int{-4, -1, 0, 3, 10}
	fmt.Println(sortedSquares(testNums))
}

File: 01_arrays/d04_squares_of_a_sorted_array.py
def solution(nums):
    res = [0] * len(nums)
    i, j = 0, len(nums) - 1
    n = len(nums) -1
    while i <= j:
        if abs(nums[i]) < abs(nums[j]):
            res[n] = nums[j] ** 2
            j -= 1
            n -= 1
        else:
            res[n] = nums[i] ** 2
            i += 1
            n -= 1
    return res


nums = list(map(int, input().split()))
print(solution(nums))
